Modules/signalmodule.c

```diff
@@ -1706,10 +1706,27 @@ _PyErr_CheckSignalsTstate(PyThreadState *tstate)
         }
         _Py_atomic_store_relaxed(&Handlers[i].tripped, 0);
 
+        /* Signal handlers can be modified while a signal is received,
+         * and therefore the fact that trip_signal() or PyErr_SetInterrupt()
+         * was called doesn't guarantee that there is still a Python
+         * signal handler for it by the time PyErr_CheckSignals() is called.
+         */
+        PyObject *func = Handlers[i].func;
+        if (func == NULL || func == Py_None || func == IgnoreHandler ||
+            func == DefaultHandler) {
+            /* No Python signal handler, continue
+             * XXX this means the signal isn't handled by anyone
+             * (should we call raise() here?), but it's better than
+             * raising cryptic exceptions asynchronously
+             * such as "TypeError: 'int' object is not callable".
+             */
+            continue;
```

Comment:
This should probably require some consensus, but I would personally raise, as the situation is tricky enough that I think it should not pass silently.

Reply:
@vstinner What do you think?

Reply:
One problem is that re-raising will break the assumption that `_thread.interrupt_main` only _simulates_ SIGINT.

Reply:
> One problem is that re-raising will break the assumption that `_thread.interrupt_main` only _simulates_ SIGINT.

That's an excellent point actually. Maybe we should set an unraisable exception?

Reply:
That would not be much better than the asynchronous `TypeError`, would it?

Reply:
It does not set the error indicator IIRC and can be handled separately if needed by a hook.

Reply:
Ah, sounds good then!

Reply:
Feel free to take a look at the updated patch @pablogsal .